### PR TITLE
[Serve] Maintain replica denylist in HTTP Proxies

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

When the Serve Controller is down, `ServeHandles` don't receive updates on which replicas are alive. During extended periods of controller downtime (e.g. when the head node crashes, and kills the controller and the GCS), `ServeHandles` may send requests to dead replicas. This can cause the requests to hang, limited only by a user-configured env var (`RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S`). However, the proxy will continue sending requests to these dead replicas– even if the requests keep timing out. This unnecessarily raises request latency, since some requests will hang until the timeout before they're retried at another (hopefully) live replica.

This change adds a local replica denylist to the HTTP Proxy's `ServeHandle`, allowing it to temporarily stop sending requests to replicas that don't respond before the timeout. This helps limit request tail latency during periods of controller downtime.

## Related issue number

Closes [#18321](https://github.com/anyscale/product/issues/18321)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@shrekris-anyscale btw, this denylist should apply for all `ServeHandle`s right? Not just HTTP proxies.

File: python/ray/serve/_private/constants.py [hunk 1]
```diff
@@ -46,6 +46,9 @@
 #: Max time to wait for HTTP proxy in `serve.start()`.
 HTTP_PROXY_TIMEOUT = 60
 
+#: Time that must pass before retrying an embargoed replica
+EMBARGO_TIMEOUT_S = 60
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/constants.py:50]: Do we need to handle this explicitly? If the handle gets a `RayActorError`, then the GCS is alive and the controller is alive or will be restarted soon. The controller should update the handle's replica set and remove the dead replica.

**Comment by shrekris-anyscale** [comment]
> @shrekris-anyscale btw, this denylist should apply for all `ServeHandles` right? Not just HTTP proxies.

@edoakes Yes, but not out-of-the-box. The `ServeHandle`'s caller must explicitly denylist a replica by calling the handle's new `embargo_replica()` function.

I considered a design where the `ServeHandle` transparently denylists failed replicas. However, this requires the handle itself to somehow track whether calls to replicas have failed or succeeded within some time limit. The handle would need to keep a reference to the latest call for each replica and periodically `await` them to check if they've succeeded. Currently, the handle is a pretty lightweight object, but this would increase the handle's compute and memory usage. That's why I went with the explicit denylist design.

What do you think about denylisting explicitly vs. doing it transparently?

**Comment by edoakes** [comment]
I think it's far preferable if we can do it transparently in the handle, but if it is significantly harder to do it that was I'm OK with doing this as an incremental step (keeping `embargo` as a private_api) and then transitioning it into the handle later.

**Comment by shrekris-anyscale** [comment]
> I think it's far preferable if we can do it transparently in the handle, but if it is significantly harder to do it that was I'm OK with doing this as an incremental step (keeping embargo as a private_api) and then transitioning it into the handle later.

Sounds good. I'll keep it outside the handle for now, but we can move it into the handle in the future.

**Pull request closed without merging**